lib/core/src/Cardano/Wallet/Network.hs

```diff
@@ -67,3 +75,20 @@ data ErrPostTx
     deriving (Generic, Show, Eq)
 
 instance Exception ErrPostTx
+
+-- | Tries to waits 20 s, until 'networkTip networkLayer' succeeds.
+waitForConnection
+    :: NetworkLayer IO
+    -> Int -- ^ Number of retries / timeout in seconds
+    -> IO ()
+waitForConnection nw = loop
+  where
+    loop :: Int -> IO ()
+    loop retries = runExceptT (networkTip nw) >>= \case
+        Right _ -> do
+            return ()
+        Left (ErrNetworkTipNetworkUnreachable _) | retries > 0 -> do
+                     TIO.putStrLn "[INFO] waiting for connection to the node..."
+                     threadDelay 1000000
+                     loop (retries - 1)
+        Left e -> throwIO e
```

Comment:
I'd suggest to simply use functions from `retry` here, which already encapsulate nicely this logic. This would also allow to simply take a `retryPolicy` as an argument (instead of an untyped `Int`) and, to provide a default policy next to the `waitForConnection` function for an easy usage:

```hs
waitForConnection defaultPolicy 
```

In testing, this would allow to pass in a different policy for which the cumulative delay can be just 500ms or so, to avoid hanging for nothing in tests.